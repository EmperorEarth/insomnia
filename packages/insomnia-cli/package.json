{
  "name": "insomnia-cli",
  "version": "0.0.1",
  "main": "dist/inso.js",
  "bin": {
    "inso": "dist/inso.js"
  },
  "scripts": {
    "typecheck": "",
    "test": "jest",
    "test:snapshots": "jest -u",
    "start": "npm run watch",
    "clean": "rimraf dist",
    "build": "babel ./src --out-dir=./dist --ignore '**/*/__tests__/**'",
    "watch": "babel --source-maps --watch=./src --out-dir=./dist --ignore '**/*/__tests__/**'",
    "bootstrap": "npm run build",
    "prepublish": "npm run build"
  },
  "jest": {
    "testMatch": [
      "**/__tests__/**/*.test.js"
    ],
    "verbose": false,
    "resetMocks": true,
    "resetModules": true
  },
  "devDependencies": {
    "@babel/cli": "^7.10.1",
    "@babel/core": "^7.9.0",
    "@babel/plugin-proposal-optional-chaining": "^7.9.0",
    "@babel/preset-env": "^7.10.2",
    "@babel/preset-flow": "^7.10.1",
    "execa-wrap": "^1.4.0",
    "flow-bin": "^0.126.1",
    "husky": "^3.1.0",
    "jest": "^26.0.1",
    "rimraf": "^3.0.2"
  },
  "dependencies": {
    "commander": "^5.1.0",
    "openapi-2-kong": "^2.2.6",
    "yaml": "^1.10.0"
  }
}
