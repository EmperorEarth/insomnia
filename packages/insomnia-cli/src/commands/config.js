// @flow
import { createCommand } from 'commander';

type ConfigGenerateOptions = {|
  type: string,
  output?: string,
|};

function handleGenerateConfig(opts: ConfigGenerateOptions) {
  const { type, output } = opts;
  console.log('Generate config for %s at %s', type, output);
}

export function makeConfigCommand() {
  const generate = createCommand('config').description('Generate something...');

  generate
    .command('generate')
    .requiredOption('-t, --type <type>', 'must specify type of configuration to generate')
    .option('-o, --output <dir>', 'where to output', '.')
    .description('Generate configuration')
    .action(handleGenerateConfig);

  return generate;
}
