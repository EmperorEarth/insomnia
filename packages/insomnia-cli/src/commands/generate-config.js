import React from 'react';
import { render } from 'ink';
import { Counter } from './counter';

export default () => {
  console.log('generate');
  render(<Counter />);
};
