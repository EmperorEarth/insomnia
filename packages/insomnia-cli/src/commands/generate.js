// @flow
import { createCommand } from 'commander';
import * as o2k from 'openapi-2-kong';
import YAML from 'yaml';
import path from 'path';
import fs from 'fs';

const ConversionTypeMap: { [string]: o2k.ConversionResultType } = {
  kubernetes: 'kong-for-kubernetes',
  declarative: 'kong-declarative-config',
};

type GenerateConfigOptions = {|
  type: $Keys<typeof ConversionTypeMap>,
  output?: string,
|};

async function handleGenerateConfig(
  filePath: string,
  { type, output }: GenerateConfigOptions,
): Promise<void> {
  if (!ConversionTypeMap[type]) {
    console.log(
      `--type ${type} not recognized. Options are: [${Object.keys(ConversionTypeMap).join(', ')}]`,
    );
    return;
  }

  const result = await o2k.generate(filePath, ConversionTypeMap[type]);

  const yamlDocs = result.documents.map(d => YAML.stringify(d));

  // Join the YAML docs with "---" and strip any extra newlines surrounding them
  const document = yamlDocs.join('\n---\n').replace(/\n+---\n+/g, '\n---\n');

  if (output) {
    const fullOutputPath = path.resolve(output);
    fs.writeFileSync(fullOutputPath, document);
  } else {
    console.log(document);
  }
}

export function makeGenerateCommand() {
  const generate = createCommand('generate');

  generate
    .command('config <filePath>')
    .option(
      '-t, --type <type>',
      `type of configuration - options: [${Object.keys(ConversionTypeMap).join(', ')}]`,
    )
    .option('-o, --output <name>', 'output path')
    .description('Generate configuration from an api spec')
    .action(handleGenerateConfig);

  return generate;
}
