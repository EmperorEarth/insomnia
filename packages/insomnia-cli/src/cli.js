// @flow

import { program } from 'commander';
import { makeGenerateCommand } from './commands/generate';
import * as packageJson from '../package.json';

export async function go() {
  await program
    .storeOptionsAsProperties(false)
    .passCommandToAction(false)
    .version(packageJson.version, '-v, --version')
    .addCommand(makeGenerateCommand())
    .parseAsync(process.argv);
}
