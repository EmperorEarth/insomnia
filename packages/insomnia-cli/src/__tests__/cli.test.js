import * as cli from '../cli';
import execa from 'execa-wrap';
import * as packageJson from '../../package.json';
import { generateConfig } from '../commands/generate';
jest.mock('../commands/generate');

const initInso = () => {
  return args => {
    const cliArgs = `node test ${args}`.split(' ');
    // console.log('calling cli.go with: %o', cliArgs);
    return cli.go(cliArgs);
  };
};

const insoSnapshot = async args =>
  expect(await execa('bin/inso', args.split(' '))).toMatchSnapshot();

describe('inso', () => {
  let inso = initInso();
  beforeEach(() => {
    inso = initInso();
  });

  it.each(['-h', '--help', 'help', 'generate -h', 'generate config -h'])(
    'shows help page with "%s"',
    async args => {
      await insoSnapshot(args);
    },
  );

  it.each(['-v', '--version'])('should print version from package.json - "%s"', async arg => {
    expect(await execa('bin/inso', [arg], { filter: ['stdout'] })).toContain(packageJson.version);
  });

  describe('generate config', () => {
    it('should error when required --type option is missing', async () => {
      await insoSnapshot('generate config');
    });

    it('should error when filePath is missing', async () => {
      await insoSnapshot('generate config -t declarative');
    });

    it('should call generateConfig', async () => {
      inso('generate config -t declarative file.yaml');
      expect(generateConfig).toHaveBeenCalledWith({
        filePath: 'file.yaml',
        type: 'declarative',
        output: undefined,
      });
    });
  });
});
