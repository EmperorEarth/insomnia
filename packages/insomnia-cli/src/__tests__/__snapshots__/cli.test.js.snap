// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inso base shows help page with "--help" 1`] = `
"
  command: bin/inso --help
  code: 0
  failed: false
  killed: false
  signal: null
  timedOut: false

  stdout:
  -------
  Usage: inso [options] [command]

  A CLI for Insomnia!

  Options:
    -v, --version   output the version number
    -h, --help      display help for command

  Commands:
    generate        Code generation utilities
    help [command]  display help for command
  -------
  stderr:
  -------
  
  -------
  "
`;

exports[`inso base shows help page with "-h" 1`] = `
"
  command: bin/inso -h
  code: 0
  failed: false
  killed: false
  signal: null
  timedOut: false

  stdout:
  -------
  Usage: inso [options] [command]

  A CLI for Insomnia!

  Options:
    -v, --version   output the version number
    -h, --help      display help for command

  Commands:
    generate        Code generation utilities
    help [command]  display help for command
  -------
  stderr:
  -------
  
  -------
  "
`;

exports[`inso base shows help page with "generate -h" 1`] = `
"
  command: bin/inso generate -h
  code: 0
  failed: false
  killed: false
  signal: null
  timedOut: false

  stdout:
  -------
  Usage: inso generate [options] [command]

  Code generation utilities

  Options:
    -h, --help                   display help for command

  Commands:
    config [options] <filePath>  Generate configuration from an api spec
    help [command]               display help for command
  -------
  stderr:
  -------
  
  -------
  "
`;

exports[`inso base shows help page with "generate config -h" 1`] = `
"
  command: bin/inso generate config -h
  code: 0
  failed: false
  killed: false
  signal: null
  timedOut: false

  stdout:
  -------
  Usage: inso generate config [options] <filePath>

  Generate configuration from an api spec

  Options:
    -t, --type <value>   the type of configuration to generate, options are
                         [kubernetes, declarative]
    -o, --output <path>  the output path
    -h, --help           display help for command
  -------
  stderr:
  -------
  
  -------
  "
`;

exports[`inso base shows help page with "help" 1`] = `
"
  command: bin/inso help
  code: 0
  failed: false
  killed: false
  signal: null
  timedOut: false

  stdout:
  -------
  Usage: inso [options] [command]

  A CLI for Insomnia!

  Options:
    -v, --version   output the version number
    -h, --help      display help for command

  Commands:
    generate        Code generation utilities
    help [command]  display help for command
  -------
  stderr:
  -------
  
  -------
  "
`;

exports[`inso generate config should error when filePath is missing 1`] = `
"
  command: bin/inso generate config -t declarative
  code: 1
  failed: true
  killed: false
  signal: null
  timedOut: false

  stdout:
  -------
  
  -------
  stderr:
  -------
  error: missing required argument 'filePath'
  -------
  "
`;

exports[`inso generate config should error when required --type option is missing 1`] = `
"
  command: bin/inso generate config
  code: 1
  failed: true
  killed: false
  signal: null
  timedOut: false

  stdout:
  -------
  
  -------
  stderr:
  -------
  error: required option '-t, --type <value>' not specified
  -------
  "
`;
